maven-core/src/main/java/org/apache/maven/execution/DefaultBuildResumptionManager.java

```diff
@@ -0,0 +1,308 @@
+        }
+    }
+
+    @Override
+    public String getResumeFromSelector( List<MavenProject> mavenProjects, MavenProject failedProject )
+    {
+        boolean hasOverlappingArtifactId = mavenProjects.stream()
+                .filter( project -> failedProject.getArtifactId().equals( project.getArtifactId() ) )
+                .count() > 1;
+
+        if ( hasOverlappingArtifactId )
+        {
+            return failedProject.getGroupId() + ":" + failedProject.getArtifactId();
+        }
+
+        return ":" + failedProject.getArtifactId();
+    }
+
+    @VisibleForTesting
+    Properties determineResumptionProperties( MavenExecutionResult result )
+    {
+        Properties properties = new Properties();
+
+        List<MavenProject> failedProjects = getFailedProjectsInOrder( result );
+        if ( !failedProjects.isEmpty() )
+        {
+            MavenProject resumeFromProject = failedProjects.get( 0 );
+            Optional<String> resumeFrom = getResumeFrom( result, resumeFromProject );
+            Optional<String> projectsToSkip = determineProjectsToSkip( result, failedProjects, resumeFromProject );
+
+            resumeFrom.ifPresent( value -> properties.setProperty( RESUME_FROM_PROPERTY, value ) );
+            projectsToSkip.ifPresent( value -> properties.setProperty( EXCLUDED_PROJECTS_PROPERTY, value ) );
+        }
+        else
+        {
+            logger.warn( "Could not create " + RESUME_PROPERTIES_FILENAME + " file: no failed projects found" );
+        }
+
+        return properties;
+    }
+
+    private List<MavenProject> getFailedProjectsInOrder( MavenExecutionResult result )
+    {
+        List<MavenProject> sortedProjects = result.getTopologicallySortedProjects();
+
+        return result.getExceptions().stream()
+                .filter( LifecycleExecutionException.class::isInstance )
+                .map( LifecycleExecutionException.class::cast )
+                .map( LifecycleExecutionException::getProject )
+                .sorted( comparing( sortedProjects::indexOf ) )
+                .collect( Collectors.toList() );
+    }
+
+    /**
+     * Determine the project where the next build can be resumed from.
+     * If the failed project is the first project of the build,
+     * it does not make sense to use --resume-from, so the result will be empty.
+     * @param result The result of the Maven build.
+     * @param failedProject The first failed project of the build.
+     * @return An optional containing the resume-from suggestion.
+     */
+    private Optional<String> getResumeFrom( MavenExecutionResult result, MavenProject failedProject )
+    {
+        List<MavenProject> allSortedProjects = result.getTopologicallySortedProjects();
+        if ( !allSortedProjects.get( 0 ).equals( failedProject ) )
+        {
+            return Optional.of( String.format( "%s:%s", failedProject.getGroupId(), failedProject.getArtifactId() ) );
```

Comment:
Inconsistent with `getResumeFromSelector()`. I see no real benefit of `format()` for this case. Static concat will be replaced with a builder at compile time anyway.

Reply:
I'm fine to replace format with string concat, I don't think it matters much in the sake of readability. 

The inconsistency of `getResumeFromSelector()` is on purpose. I think that method was designed in order for the user to get the most short-hand hint to "resume-from" the build with. In this case the user does not have to see this selector at all, so we can use the full unique key (group+artifact id). The added benefit is that we do not have to calculate whether an artifact-id alone is unique in the whole project.